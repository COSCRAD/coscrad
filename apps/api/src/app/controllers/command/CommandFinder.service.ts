import { Injectable } from '@nestjs/common';
import { ModulesContainer } from '@nestjs/core';
import { COMMAND_HANDLER_METADATA } from './decorators/CommandHandler.decorator';
import { ICommandHandler } from './ICommandHandler';

@Injectable()
export class CommandFinderService {
    constructor(private readonly modulesContainer: ModulesContainer) {}

    explore(): ICommandHandler[] {
        const modules = [...this.modulesContainer.values()];

        const result = modules
            .map((module) => [...module.providers.values()])
            .reduce((a, b) => a.concat(b), [])
            .filter((element) => !!element)
            .map(({ instance }) => instance)
            .filter((instance) => !!instance)
            .filter((instance) => !!instance.constructor)
            .map((instance) => instance.constructor)
            .map((constructor) => Reflect.getMetadata(COMMAND_HANDLER_METADATA, constructor))
            .filter((meta) => !!meta);

        console.log({
            foundCommandHandlers: result,
        });

        return result;
    }
}
