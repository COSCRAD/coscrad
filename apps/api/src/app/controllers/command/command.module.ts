import { Module, OnApplicationBootstrap } from '@nestjs/common';
import { CommandFinderService } from './CommandFinder.service';
import CommandHandlerService from './CommandHandler.service';

@Module({
    providers: [CommandHandlerService, CommandFinderService],
    exports: [CommandHandlerService],
})
export class CommandModule implements OnApplicationBootstrap {
    constructor(
        private readonly finderService: CommandFinderService,
        private readonly commandHanlderService: CommandHandlerService
    ) {}

    onApplicationBootstrap() {
        const commandHandlers = this.finderService.explore();

        commandHandlers.forEach((handler) => this.commandHanlderService.registerHandler(handler));

        console.log({
            commandHandlerService: JSON.stringify(this.commandHanlderService),
        });
    }
}
