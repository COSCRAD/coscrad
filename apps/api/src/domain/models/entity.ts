import { PartialDTO } from 'apps/api/src/types/partial-dto';
import { EntityType } from '../types/entityType';
import { isNullOrUndefined } from '../utilities/validation/is-null-or-undefined';
import { EntityId } from './types/entity-id';
import { EntityCompositeIdentifier } from './types/entityCompositeIdentifier';

// TODO Extract to a separate location
const isValidStringWithLength = (input: unknown): input is string =>
  typeof input === 'string' && input.length > 0;

/**
 *
 * TODO move this near the corresponding type. Note that this extra layer is
 * meant to safe guard us in case we want stricter validation on an `entity id`.
 * Should we change the rule, we can update this in once place, instead of replacing
 * `isValidStringWithLength` everywhere.
 */
const isValidEntityId = (input: unknown): input is EntityId =>
  isValidStringWithLength(input);

export abstract class Entity {
  readonly id: EntityId;

  abstract readonly type: EntityType;

  readonly published: boolean;

  constructor(dto: unknown) {
    // TODO Validate DTO
    if (!this.#validateDTO(dto))
      throw new Error(`Invalid dto for an entity: ${JSON.stringify(dto)}`);

    this.id = dto.id;

    this.published = typeof dto.published === 'boolean' ? dto.published : false;
  }

  getCompositeIdentifier = (): EntityCompositeIdentifier => ({
    type: this.type,
    id: this.id,
  });

  /**
   * TODO consider moving this to a `Serializable` mixin. There may be cases
   * where we want this behaviour on a non domain-model class without the
   * inheritance baggage.
   *  */
  toDTO<TEntity extends Entity>(this: TEntity): PartialDTO<TEntity> {
    return JSON.parse(JSON.stringify(this));
  }

  /**
   * TODO develop a general pattern \ directory structure for validation \ validators.
   * One idea: Return a symbol Valid | Errors[] from a validator instead of a
   * boolean flag.
   *
   * We may break the validators out into separate files. Then we will validate
   * a DTO before passing it to the constructor in our `instance factory` methods
   * in the mapping layer for the repositories.
   *  */
  #validateDTO = (dto: unknown): dto is PartialDTO<Entity> => {
    const { id } = dto as PartialDTO<Entity>;

    // A new entity will have id auto assigned by db
    if (!isNullOrUndefined(id) && !isValidEntityId(id)) return false;

    return true;
  };
}
