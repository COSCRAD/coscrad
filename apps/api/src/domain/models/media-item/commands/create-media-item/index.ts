export * from './create-media-item.command';
export * from './create-media-item.command-handler';
export * from './media-item-created.event';
