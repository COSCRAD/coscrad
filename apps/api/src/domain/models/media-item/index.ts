export * from './media-item.module';
