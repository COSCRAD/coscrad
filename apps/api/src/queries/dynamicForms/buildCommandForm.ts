import {
    AGGREGATE_COMPOSITE_IDENTIFIER,
    CoscradComplexDataType,
    CoscradDataType,
    ICoscradModelSchema,
    IDynamicForm,
    IFormField,
} from '@coscrad/api-interfaces';
import {
    CommandContext,
    isCommandWriteContext,
} from '../../app/controllers/command/services/command-info-service';
import { buildFormFieldForCommandPayloadProp } from './buildFormFieldForCommandPayloadProp';

// TODO We should have a SST for `rawData` as a key on all CREATE_X payloads.
const propertyKeysToOmitFromForms = [AGGREGATE_COMPOSITE_IDENTIFIER, 'rawData'];

const isPropertyOmittedFromForm = (key, propertySchema) => {
    if (propertyKeysToOmitFromForms.includes(key)) return true;

    return propertySchema?.complexDataType === CoscradComplexDataType.union;
};

export const buildCommandForm = <T extends Record<string, unknown>>(
    schema: ICoscradModelSchema<T, CoscradDataType>,
    context: CommandContext
): IDynamicForm => {
    const prepopulatedFields = isCommandWriteContext(context)
        ? {
              [AGGREGATE_COMPOSITE_IDENTIFIER]: context.getCompositeIdentifier(),
          }
        : {};

    const fields: IFormField[] = Object.entries(schema).reduce(
        (acc: IFormField[], [key, propertySchema]) =>
            isPropertyOmittedFromForm(key, schema)
                ? acc
                : [
                      ...acc,
                      buildFormFieldForCommandPayloadProp(propertySchema, {
                          name: key,
                          label: propertySchema.label,
                          description: propertySchema.description,
                      }),
                  ],
        []
    );

    return {
        fields,
        prepopulatedFields,
    };
};
