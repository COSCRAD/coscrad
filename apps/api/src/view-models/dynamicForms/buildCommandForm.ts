import {
    CoscradDataType,
    ICoscradModelSchema,
    IDynamicForm,
    IFormField,
} from '@coscrad/api-interfaces';
import { buildFormFieldForCommandPayloadProp } from './buildFormFieldForCommandPayloadProp';

export const buildCommandForm = <T extends Record<string, unknown>>(
    schema: ICoscradModelSchema<T, CoscradDataType>
): IDynamicForm => {
    const fields: IFormField[] = Object.entries(schema).reduce(
        (acc: IFormField[], [key, propertySchema]) => [
            ...acc,
            buildFormFieldForCommandPayloadProp(propertySchema, {
                name: key,
                /**
                 * TODO [https://www.pivotaltracker.com/story/show/183504291]
                 * We need decorators for command prop labels and descriptions.
                 */
                label: 'add label',
                description: 'add description',
            }),
        ],
        []
    );

    return {
        fields,
    };
};
