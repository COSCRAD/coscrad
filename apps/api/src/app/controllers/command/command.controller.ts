import { AGGREGATE_COMPOSITE_IDENTIFIER } from '@coscrad/api-interfaces';
import { Ack, CommandHandlerService } from '@coscrad/commands';
import {
    Body,
    Controller,
    MessageEvent,
    Post,
    Request,
    Res,
    Sse,
    UnauthorizedException,
    UseFilters,
    UseGuards,
} from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
import { Observable, Subject } from 'rxjs';
import { CoscradUserWithGroups } from '../../../domain/models/user-management/user/entities/user/coscrad-user-with-groups';
import httpStatusCodes from '../../constants/httpStatusCodes';
import { CommandWithGivenTypeNotFoundExceptionFilter } from '../exception-handling/exception-filters/command-with-given-type-not-found.filter';
import { NoCommandHandlerForCommandTypeFilter } from '../exception-handling/exception-filters/no-command-handler-for-command-type.filter';
import sendInternalResultAsHttpResponse from '../resources/common/sendInternalResultAsHttpResponse';
import { CommandFSA } from './command-fsa/command-fsa.entity';

export const AdminJwtGuard = AuthGuard('jwt');

@ApiTags('commands')
@Controller('commands')
/**
 * TODO [https://www.pivotaltracker.com/story/show/182785593]
 * Enforce RBAC for commands in a guard isntead.
 */
/**
 * The next two filters convert a thrown error to a returned error (400) when an
 * invalid command type is provided by the user.
 *
 * TODO [https://www.pivotaltracker.com/story/show/182785593]
 * We may want to do this in a pipe in the future.
 */
@UseFilters(new CommandWithGivenTypeNotFoundExceptionFilter())
@UseFilters(new NoCommandHandlerForCommandTypeFilter())
export class CommandController {
    private readonly commandResultSubject = new Subject<MessageEvent>();

    constructor(private readonly commandHandlerService: CommandHandlerService) {}

    @ApiBearerAuth('JWT')
    @UseGuards(AdminJwtGuard)
    @Post('')
    async executeCommand(@Request() req, @Res() res, @Body() commandFSA: CommandFSA) {
        const { user } = req;

        if (!user || !(user instanceof CoscradUserWithGroups)) {
            throw new UnauthorizedException();
        }

        if (!user.isAdmin()) {
            throw new UnauthorizedException();
        }

        /**
         * Note that we defer command type validation to the command handler.
         * This is because we want to keep the controller free of domain
         * logic. If we want to drive commands via a CLI, it shouldn't need
         * to know about http.
         */
        const { type, payload } = commandFSA;

        const result = await this.commandHandlerService.execute(
            { type, payload },
            { userId: user.id }
        );

        if (result !== Ack) return sendInternalResultAsHttpResponse(res, result);

        this.commandResultSubject.next({
            data: { aggregateCompositeIdentifier: payload[AGGREGATE_COMPOSITE_IDENTIFIER] },
        });

        return res.status(httpStatusCodes.ok).send();
    }

    @Sse('notifications')
    commandSuccessNotifications(): Observable<MessageEvent> {
        return this.commandResultSubject.asObservable();
    }
}
