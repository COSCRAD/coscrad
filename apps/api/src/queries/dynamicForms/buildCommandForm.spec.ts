import { getCoscradDataSchema } from '@coscrad/data-types';
import { CreateMediaItem } from '../../domain/models/media-item/commands/create-media-item/create-media-item.command';
import { MediaItem } from '../../domain/models/media-item/entities/media-item.entity';
import { buildCommandForm } from './buildCommandForm';

describe('buildCommandForm', () => {
    it('should return the expected result', () => {
        const commandSchema = getCoscradDataSchema(CreateMediaItem);

        const form = buildCommandForm(commandSchema, MediaItem);

        expect(form).toMatchSnapshot();
    });
});
