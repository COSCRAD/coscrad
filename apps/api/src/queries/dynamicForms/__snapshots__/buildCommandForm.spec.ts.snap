// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildCommandForm should return the expected result 1`] = `
{
  "fields": [
    {
      "constraints": [
        {
          "message": "must be a non-empty string",
          "name": "non-empty string",
        },
      ],
      "description": "title of the media item in the language",
      "label": "title (language)",
      "name": "title",
      "type": "TEXT_FIELD",
    },
    {
      "constraints": [
        {
          "message": "must be a URL",
          "name": "URL",
        },
        {
          "message": "must be a defined value",
          "name": "defined value",
        },
      ],
      "description": "a web URL link to a digital version of this media item for playback",
      "label": "audio link",
      "name": "url",
      "type": "TEXT_FIELD",
    },
    {
      "constraints": [
        {
          "message": "Must be a valid \${propertyLabel}",
          "name": "IS_ENUM",
        },
        {
          "message": "Required",
          "name": "defined value",
        },
      ],
      "description": "technical specification of the type of media item",
      "label": "MIME Type",
      "name": "mimeType",
      "options": [
        {
          "display": "png",
          "value": "image/png",
        },
        {
          "display": "mp3",
          "value": "audio/mpeg",
        },
        {
          "display": "wav",
          "value": "audio/x-wav",
        },
        {
          "display": "audioOgg",
          "value": "audio/ogg",
        },
        {
          "display": "mp4",
          "value": "video/mp4",
        },
        {
          "display": "videoOgg",
          "value": "video/ogg",
        },
        {
          "display": "videoWebm",
          "value": "video/webm",
        },
      ],
      "type": "STATIC_SELECT",
    },
    {
      "constraints": [
        {
          "message": "must be a non-negative number",
          "name": "non-negative number",
        },
        {
          "message": "must be a finite number",
          "name": "finite number",
        },
      ],
      "description": "the length of the media item in milliseconds (audio or video only)",
      "label": "length (ms)",
      "name": "lengthMilliseconds",
      "type": "NUMERIC_INPUT",
    },
    {
      "constraints": [
        {
          "message": "must be a non-negative number",
          "name": "non-negative number",
        },
        {
          "message": "must be a finite number",
          "name": "finite number",
        },
      ],
      "description": "the height of the media item in pixels",
      "label": "height (px)",
      "name": "heightPx",
      "type": "NUMERIC_INPUT",
    },
    {
      "constraints": [
        {
          "message": "must be a non-negative number",
          "name": "non-negative number",
        },
        {
          "message": "must be a finite number",
          "name": "finite number",
        },
      ],
      "description": "the width of the media item in pixels",
      "label": "width (px)",
      "name": "widthPx",
      "type": "NUMERIC_INPUT",
    },
  ],
  "prepopulatedFields": {},
}
`;
